File: AoC 2023/AoC 2023/Week4.swift
//
//  Week4.swift
//  AoC 2023
//
//  Created by David Gavilan Ruiz on 03/03/2024.
//

import Foundation

func day22(input: [String]) {
    // Part 1
    let jengatris = Jengatris(input: input)
    //jengatris.initialState.volume.dump()
    let (firstFall, _) = Jengatris.simulate(start: jengatris.initialState)
    //firstFall.volume.dump()
    let essentials = Jengatris.findEssentials(state: firstFall)
    //print(essentials)
    let disposableCount = jengatris.initialState.pieces.count - essentials.count
    print("There are \(disposableCount) disposable pieces")
    // Part 2
    let t1 = measure {
        let n = Jengatris.countFalls(state: firstFall, ids: essentials)
        print("\(n) bricks would fall.")
    }
    print("Took \(t1 * 1_000) ms to simulate.")
    let t2 = measure {
        Task.synchronous {
            let cfc = Jengatris.ConcurrentFallCounter(state: firstFall)
            let n = await cfc.countFalls(ids: essentials)
            print("\(n) bricks would fall.")
        }
    }
    print("Took \(t2 * 1_000) ms to simulate (async/await)")
    let t3 = measure {
        let n = Jengatris.concurrentCountFalls(state: firstFall, ids: essentials)
        print("\(n) bricks would fall.")
    }
    print("Took \(t3 * 1_000) ms to simulate (Grand Central Dispatch)")
}

File: AoC 2023/cpp/advent2023-cpp/week4.cpp
//
//  week4.cpp
//  advent2023-cpp
//
//  Created by David Gavilan Ruiz on 04/03/2024.
//

#include <stdio.h>
#include <iostream>
#include "week4.hpp"
#include "jengatris.hpp"
#include "util.hpp"

namespace advent
{

void day22(std::istream& input)
{
    auto jenga = Jengatris(input);
    auto gameState = Jengatris::simulate(jenga.getState());
    auto essentials = Jengatris::findEssentials(*gameState);
    auto disposableCount = gameState->pieces.size() - essentials.size();
    std::cout << "There are " << disposableCount << " disposable pieces." << std::endl;
    // Part 2
    size_t n;
    auto elapsed = measure([essentials, gameState, &n]() {
        n = Jengatris::countFalls(*gameState, essentials);
    });
    std::cout << n << " bricks would fall. Took " << elapsed << " ms." << std::endl;
    elapsed = measure([essentials, gameState, &n]() {
        n = Jengatris::countFallsThreaded(*gameState, essentials);
    });
    std::cout << n << " bricks would fall. Took " << elapsed << " ms. (threads)" << std::endl;
    elapsed = measure([essentials, gameState, &n]() {
        n = Jengatris::countFallsAsync(*gameState, essentials);
    });
    std::cout << n << " bricks would fall. Took " << elapsed << " ms. (async)" << std::endl;
#ifndef __clang__
    // To test these, install TBB and compile with gcc, i.e.
    // > brew install tbb
    // > g++-13 -std=c++17 -O3 -Wall -Wextra -pedantic -o advent advent2023-cpp/*.cpp -ltbb -I/opt/homebrew/Cellar/tbb/2021.11.0/include/ -L/opt/homebrew/Cellar/tbb/2021.11.0/lib
    // > ./advent ../Resources/day22_input.txt
    elapsed = measure([essentials, gameState, &n]() {
        n = Jengatris::countFallsParallel(*gameState, essentials);
    });
    std::cout << n << " bricks would fall. Took " << elapsed << " ms. (parallel)" << std::endl;
    elapsed = measure([essentials, gameState, &n]() {
        n = Jengatris::countFallsTBB(*gameState, essentials);
    });
    std::cout << n << " bricks would fall. Took " << elapsed << " ms. (TBB)" << std::endl;
#endif
}

}
